File: 3217-delete-nodes-from-linked-list-present-in-array/main.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* modifiedList(vector<int>& nums, ListNode* head) {

        if (!head) return head;

        std::unordered_set<int> numSet(nums.begin(), nums.end());

        ListNode* curr = head;
        ListNode* prev = nullptr;

        while (curr) {
            if (numSet.find(curr->val) != numSet.end()) {
                if (prev) {
                    prev->next = curr->next;
                } else {
                    head = curr->next;
                }
            } else {
                prev = curr;
            }
            curr = curr->next;
        }

        return head;    
    }
};

File: 3217-delete-nodes-from-linked-list-present-in-array/main.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def modifiedList(self, nums: List[int], head: Optional[ListNode]) -> Optional[ListNode]:
        
        if not head: return head

        curr = head
        prev = None

        nums = set(nums)

        while curr:
            if curr.val in nums:
                if prev:
                    prev.next = curr.next
                else:
                    head = curr.next
            else:
                prev = curr

            curr = curr.next

        return head
